labs/b9.md

```diff
@@ -0,0 +1,178 @@
+
+>Unable to find image 'hello-world:latest' locally  
+latest: Pulling from library/hello-world  b04784fba78d: Already exists  
+Digest: sha256:f3b3b28a45160805bb16542c9531888519430e9e6d6ffc09d72261b0d26ff74f
+Status: Downloaded newer image for hello-world:latest  Hello from Docker!    
+
+This message shows that your installation appears to be working correctly. To generate this message, Docker took the following steps:    
+
+1. The Docker client contacted the Docker daemon.  
+2. The Docker daemon pulled the "hello-world" image from the Docker Hub.    
+3. The Docker daemon created a new container from that image which runs the executable that produces the output you are currently reading.    
+ 4. The Docker daemon streamed that output to the Docker client, which sent it to your terminal.    
+
+Some [quick definitions from Docker’s website](https://docs.docker.com/get-started/#a-brief-explanation-of-containers):    
+
+An `image` is a lightweight, stand-alone, executable package that includes everything needed to run a piece of software, including the code, a runtime, libraries, environment variables, and config files.        
+
+A `container` is a runtime instance of an image—what the image becomes in memory when actually executed. It runs completely isolated from the host environment by default, only accessing host files and ports if configured to do so.        
+
+Be sure to read through the output from running the hello-world image to get an understanding of what the Docker daemon was doing.    
+
+#### Running an interactive container    
+
+We’re now going to walk you through running a container interactively. This is useful if you ever need to play around and install stuff on a bare system without messing up your current system. Try running the following command:    
+
+`docker run -it ubuntu:latest`    
+
+The -i flag tells docker to keep `STDIN` open to your container, and the -t flag allocates a [pseudo-TTY](https://en.wikipedia.org/wiki/Pseudoterminal) for you. Basically you need both for you to have a shell into your newly started container. Try installing some packages from apt or just play around. It should look like a bare Linux system.    
+
+You can exit the container with CTRL+D.    
+
+##### Questions    
+
+1.  What user are you logged in as by default?    
+2.  If you start and then exit an interactive container, and then use the `docker run -it ubuntu:latest` command again; is it the same container? How can you tell?      
+
+#### Dockerfiles    
+
+A more powerful way to interface with Docker is by using a `Dockerfile`. A Dockerfile allows you to define an image by specifying all of the commands you would type manually to create an image. Docker can then build images from a specified Dockerfile. These Dockerfiles can be put into version control and the images distributed as a binary to keep track of both how the image is constructed and also to keep pre-built images around.        
```

Comment:
I'd probably say here that the pre-built images are useful because of speed primarily, but also because you can build something in the past with some content and that a current build could create different content because of a different base image pulled, etc.

Reply:
This is good, but seems more I'm gonna add it to the definitions above.